// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
edition = "2023";

// This proto is designed to support research on AI fitness coaches.

package phllm_protos;

import "third_party/tensorflow/core/example/feature.proto";

option features.utf8_validation = NONE;

// User supplied inputs relevant to the case study.
message UserInput {
  // 'Tags' for user interactions; a short name for this interaction.
  // Example conceived tag: 'ask-user-for-sleep-perception'
  // This is effectively a loosely typed enum to provide flexibility downstream.
  string tag = 1;

  // The prompt shown to the user to collect this feedback. This may not
  // necessarily exist if the user provided data unprompted.
  string prompt = 2;

  // The response provided by the user (potentially freeform)
  string response = 3;
}

// Features associated with this case study. They can be tagged for namespacing.
// We use TF Features to allow us to encode a dictionary of tensors, allowing
// arbitrary data to be placed inside a feature.
message Feature {
  // A tag for this feature set: e.g. 'daily_summaries'
  string tag = 1;

  // This is just a map of string -> tf.Features.
  // No explicit timestamp field has been provided. It is assumed that this will
  // be embedded into the features.
  tensorflow.Features tensors = 2;
}

// Top level case study data structure.
message CaseStudy {
  // An identifier for this case study.
  string case_study_id = 1;

  // Additional features available for this case study. This can include:
  // Contextual data, or time series.
  repeated Feature contextual_features = 2;

  // User inputs associated with this case study.
  repeated UserInput user_inputs = 3;
}
